File: NextApp/Android/core/model/src/main/kotlin/jp/komehyappyo/member/next/core/model/Post.kt
package jp.komehyappyo.member.next.core.model

data class PostAttachment(
    val type: String,
    val name: String,
    val url: String,
)

data class AdminReply(
    val id: String,
    val postId: String,
    val adminUserId: String,
    val adminName: String,
    val body: String,
    val createdAt: String?,
)

data class MemberPost(
    val id: String,
    val communityId: String,
    val authorUserId: String,
    val authorName: String,
    val title: String,
    val body: String,
    val attachments: List<PostAttachment> = emptyList(),
    val status: String = "submitted",
    val adminReply: String = "",
    val memberHasReadReply: Boolean = true,
    val createdAt: String? = null,
    val updatedAt: String? = null,
) {
    val hasUnreadReply: Boolean
        get() = adminReply.isNotBlank() && !memberHasReadReply

    fun canEdit(userId: String): Boolean = authorUserId == userId
}

data class PublicPost(
    val id: String,
    val authorUserId: String,
    val authorName: String,
    val title: String,
    val body: String,
    val categoryId: String?,
    val attachments: List<PostAttachment> = emptyList(),
    val createdAt: String? = null,
)

data class RadioProgram(
    val id: String,
    val communityId: String,
    val title: String,
    val description: String,
    val imageUrl: String,
    val audioUrl: String,
    val broadcastStartAt: java.time.Instant,
    val broadcastEndAt: java.time.Instant,
)

object RadioPlaybackPolicy {
    fun isPlayable(program: RadioProgram, at: java.time.Instant): Boolean =
        !at.isBefore(program.broadcastStartAt)
}

data class RadioPlaybackRecord(
    val userId: String,
    val programId: String,
    val lastPositionSeconds: Long = 0,
    val playCount: Int = 0,
    val lastPlayedAt: java.time.Instant? = null,
)

object RadioPlaybackRecordPolicy {
    fun started(
        existing: RadioPlaybackRecord?,
        userId: String,
        programId: String,
        at: java.time.Instant,
    ): RadioPlaybackRecord = RadioPlaybackRecord(
        userId = userId,
        programId = programId,
        lastPositionSeconds = maxOf(0, existing?.lastPositionSeconds ?: 0),
        playCount = (existing?.playCount ?: 0) + 1,
        lastPlayedAt = at,
    )

    fun updatingPosition(
        existing: RadioPlaybackRecord,
        positionSeconds: Long,
        at: java.time.Instant,
    ): RadioPlaybackRecord = existing.copy(
        lastPositionSeconds = maxOf(0, positionSeconds),
        lastPlayedAt = at,
    )
}

object RadioPlaybackInterruptionPolicy {
    fun shouldPause(focusChange: Int): Boolean = focusChange < 0

    fun shouldResume(wasPlayingBeforeTransientLoss: Boolean, focusChange: Int): Boolean =
        wasPlayingBeforeTransientLoss && focusChange > 0
}

object RadioPlaybackPresentation {
    const val PLAY_ACTION = "再生"
    const val PAUSE_ACTION = "一時停止"
    const val RESUME_ACTION = "再開"
    const val STOP_ACTION = "停止"
    const val PLAYING_STATUS = "再生中"
    const val PAUSED_STATUS = "一時停止中"

    fun primaryAction(isPlayable: Boolean, isActive: Boolean, isPlaying: Boolean): String = when {
        !isPlayable -> "配信前"
        isActive && isPlaying -> PAUSE_ACTION
        isActive -> RESUME_ACTION
        else -> PLAY_ACTION
    }

    fun status(isPlaying: Boolean): String = if (isPlaying) PLAYING_STATUS else PAUSED_STATUS
}

File: NextApp/iOS/OrgPortalNext/Core/Model/Post.swift
import Foundation

public struct PostAttachment: Equatable, Codable, Sendable {
    public let type: String
    public let name: String
    public let url: URL

    public init(type: String, name: String, url: URL) {
        self.type = type
        self.name = name
        self.url = url
    }
}

public struct AdminReply: Identifiable, Equatable, Codable, Sendable {
    public let id: String
    public let postId: String
    public let adminUserId: String
    public let adminName: String?
    public let body: String
    public let createdAt: Date?

    public init(
        id: String,
        postId: String,
        adminUserId: String,
        adminName: String?,
        body: String,
        createdAt: Date?
    ) {
        self.id = id
        self.postId = postId
        self.adminUserId = adminUserId
        self.adminName = adminName
        self.body = body
        self.createdAt = createdAt
    }
}

public struct MemberPost: Identifiable, Equatable, Codable, Sendable {
    public let id: String
    public let communityId: String
    public let authorUserId: String
    public let authorName: String
    public let title: String
    public let body: String
    public let attachments: [PostAttachment]
    public let status: String
    public let legacyAdminReply: String?
    public let memberHasReadReply: Bool
    public let createdAt: Date?
    public let updatedAt: Date?

    public init(
        id: String,
        communityId: String,
        authorUserId: String,
        authorName: String,
        title: String,
        body: String,
        attachments: [PostAttachment] = [],
        status: String = "submitted",
        legacyAdminReply: String? = nil,
        memberHasReadReply: Bool = true,
        createdAt: Date? = nil,
        updatedAt: Date? = nil
    ) {
        self.id = id
        self.communityId = communityId
        self.authorUserId = authorUserId
        self.authorName = authorName
        self.title = title
        self.body = body
        self.attachments = attachments
        self.status = status
        self.legacyAdminReply = legacyAdminReply
        self.memberHasReadReply = memberHasReadReply
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    public var hasUnreadReply: Bool {
        !(legacyAdminReply ?? "").isEmpty && !memberHasReadReply
    }

    public func canEdit(userId: String) -> Bool {
        authorUserId == userId
    }
}

public struct PublicPost: Identifiable, Equatable, Codable, Sendable {
    public let id: String
    public let authorUserId: String
    public let authorName: String
    public let title: String
    public let categoryId: String?
    public let body: String
    public let attachments: [PostAttachment]
    public let createdAt: Date?

    public init(
        id: String,
        authorUserId: String,
        authorName: String,
        title: String,
        categoryId: String?,
        body: String,
        attachments: [PostAttachment] = [],
        createdAt: Date? = nil
    ) {
        self.id = id
        self.authorUserId = authorUserId
        self.authorName = authorName
        self.title = title
        self.categoryId = categoryId
        self.body = body
        self.attachments = attachments
        self.createdAt = createdAt
    }
}

public struct RadioProgram: Identifiable, Equatable, Codable, Sendable {
    public let id: String
    public let communityId: String
    public let title: String
    public let description: String
    public let imageUrl: String
    public let audioUrl: URL
    public let broadcastStartAt: Date
    public let broadcastEndAt: Date

    public init(
        id: String,
        communityId: String,
        title: String,
        description: String,
        imageUrl: String,
        audioUrl: URL,
        broadcastStartAt: Date,
        broadcastEndAt: Date
    ) {
        self.id = id
        self.communityId = communityId
        self.title = title
        self.description = description
        self.imageUrl = imageUrl
        self.audioUrl = audioUrl
        self.broadcastStartAt = broadcastStartAt
        self.broadcastEndAt = broadcastEndAt
    }
}

public enum RadioPlaybackPolicy {
    public static func isPlayable(_ program: RadioProgram, at date: Date) -> Bool {
        date >= program.broadcastStartAt
    }
}

public struct RadioPlaybackRecord: Identifiable, Equatable, Codable, Sendable {
    public let id: String
    public let userId: String
    public let programId: String
    public let lastPositionSeconds: Double
    public let playCount: Int
    public let lastPlayedAt: Date?

    public init(
        id: String = UUID().uuidString,
        userId: String,
        programId: String,
        lastPositionSeconds: Double = 0,
        playCount: Int = 0,
        lastPlayedAt: Date? = nil
    ) {
        self.id = id
        self.userId = userId
        self.programId = programId
        self.lastPositionSeconds = lastPositionSeconds
        self.playCount = playCount
        self.lastPlayedAt = lastPlayedAt
    }
}

public enum RadioPlaybackRecordPolicy {
    public static func started(
        existing: RadioPlaybackRecord?,
        userId: String,
        programId: String,
        at date: Date
    ) -> RadioPlaybackRecord {
        RadioPlaybackRecord(
            id: existing?.id ?? UUID().uuidString,
            userId: userId,
            programId: programId,
            lastPositionSeconds: max(0, existing?.lastPositionSeconds ?? 0),
            playCount: (existing?.playCount ?? 0) + 1,
            lastPlayedAt: date
        )
    }

    public static func updatingPosition(
        _ existing: RadioPlaybackRecord,
        positionSeconds: Double,
        at date: Date
    ) -> RadioPlaybackRecord {
        RadioPlaybackRecord(
            id: existing.id,
            userId: existing.userId,
            programId: existing.programId,
            lastPositionSeconds: max(0, positionSeconds.isFinite ? positionSeconds : 0),
            playCount: existing.playCount,
            lastPlayedAt: date
        )
    }
}

public enum RadioPlaybackInterruptionPolicy {
    public static func shouldResume(
        wasPlayingBeforeInterruption: Bool,
        systemAllowsResume: Bool
    ) -> Bool {
        wasPlayingBeforeInterruption && systemAllowsResume
    }
}

public enum RadioPlaybackPresentation {
    public static let playAction = "再生"
    public static let pauseAction = "一時停止"
    public static let resumeAction = "再開"
    public static let stopAction = "停止"
    public static let playingStatus = "再生中"
    public static let pausedStatus = "一時停止中"

    public static func primaryAction(
        isPlayable: Bool,
        isActive: Bool,
        isPlaying: Bool
    ) -> String {
        if !isPlayable { return "配信前" }
        if isActive && isPlaying { return pauseAction }
        if isActive { return resumeAction }
        return playAction
    }

    public static func status(isPlaying: Bool) -> String {
        isPlaying ? playingStatus : pausedStatus
    }
}
